.github/workflows/debops-ci

```diff
@@ -0,0 +1,563 @@
+        filename = extra[0]
+        assert filename.endswith(".deb")
+        assert args.repo, "Please specify a repository"
+        url = f"{BINTRAY_API}/content/{args.org}/{args.repo}/{filename}"
+        resp = requests.delete(url, auth=auth)
+        if not resp.ok:
+            print(f"Error {resp.text} when calling {resp.request.url}")
+            sys.exit(1)
+
+
+class S3:
+    """S3 backend"""
+
+    def generate_release_file(self, conf, sha, size):
+        r = dedent(
+            f"""
+            Acquire-By-Hash: no
+            Architectures: {conf.arch}
+            Codename: {conf.distro}
+            Components: main
+            Date: Thu, 07 Nov 2019 14:23:37 UTC
+            Origin: private
+            Valid-Until: Thu, 14 Nov 2029 14:23:37 UTC
+            SHA256:
+             {sha} {size} main/binary-{conf.arch}/Packages
+        """
+        )
+        return r
+
+    def init_archive(self, conf):
+        """Initialize the archive"""
+        assert conf.bucket_name
+        r, o = runc(f"s3cmd mb s3://{conf.bucket_name}")
+        if r == 0:
+            print("Bucket created")
+            runc(f"s3cmd ws-create s3://{conf.bucket_name}")
+
+        r, out = runc(f"s3cmd ws-info s3://{conf.bucket_name}")
+        for li in out.splitlines():
+            if li.startswith("Website endpoint"):
+                s3url = li.split()[2]
+                break
+
+        # Initialize distro if needed. Check for InRelease
+        baseuri = f"s3://{conf.bucket_name}/dists/{conf.distro}"
+        r, o = runc(f"s3cmd info --no-progress {baseuri}/InRelease")
+        if r == 0:
+            return
+
+        if r != 12:
+            print(f"Unexpected return code {r} {o}")
+            sys.exit(1)
+
+        # InRelease file not found: create lock file
+        print("Creating initial lock file")
+        tf = NamedTemporaryFile()
+        # put = "s3cmd --acl-public --guess-mime-type --no-progress put"
+        put = "s3cmd --guess-mime-type --no-progress put"
+        r2, o = runc(f"{put} --no-progress {tf.name} {baseuri}/.debrepos3.lock")
+        assert r2 == 0, repr(o)
+
+        # Create empty InRelease
+        r2, o = runc(f"{put} {tf.name} {baseuri}/InRelease")
+
+        # Create empty Packages
+        r, o = runc(f"{put} {tf.name} {baseuri}/main/binary-{conf.arch}/Packages")
+
+        # Create index
+        html = dedent(
+            f"""
+            <html><body>
+            <p>Create /etc/apt/sources.list.d/{conf.distro}.list containing:</p>
+            <pre>deb {s3url} {conf.distro} main</pre>
+            </body></html>
+        """
+        )
+        with open(tf.name, "w") as f:
+            f.write(html)
+
+        r, o = runc(f"{put} {tf.name} {baseuri}/index.html")
+
+    def lock(self, conf, baseuri):
+        """Rename semaphore file"""
+        print(f"Locking {baseuri} ...")
+        cmd = f"s3cmd mv --no-progress {baseuri}/.debrepos3.nolock {baseuri}/.debrepos3.lock"
```

Comment:
Be careful that the mv operation is not atomic on s3 and is basically equivalent to doing a copy and delete. It might still be enough to prevent race conditions on the CI which has high latency, but it's probably worth noting this caveat in the code to prevent surprises down the line.

Reply:
The copy + delete process should implement something similar to a 2-step lock: during the interval when both files exist other attempt at copy should fail.